### PR TITLE
Add fetch() interface and docs

### PR DESCRIPTION
**Pull request by bmbouter**
- add the fetch() interface
- Docs improvements for asyncio downloaders
- fix some examples that were wrong

### CODE REVIEW

File: docs/plugins/plugin-api/asyncio.rst [hunk 1]
```diff
@@ -89,46 +178,59 @@ in parallel. See the examples below.
...
+.. _downloader-factory:
 
-Exception Handling
-------------------
+Download Factory
+----------------
+
+The DownloaderFactory constructs and configures a downloader for any given url. Specifically:
+
+1. Select the appropriate downloader based from these supported schemes: `https`, `https` or `file`.
```
Comment (dkliban) [inline docs/plugins/plugin-api/asyncio.rst]: s/https/http/

File: docs/plugins/plugin-api/asyncio.rst [hunk 2]
```diff
@@ -59,6 +61,91 @@ the downloader's `run()` method when the download is complete.
...
+busy or if a redirect occurs.
+
+Unrecoverable errors of several types can be raised during downloading. One example is a
+:ref:`validation exception <validation-exceptions>` that are raised if the content downloaded fails
+size or digest validation. There can also be protocol specific errors such as an
+``aiohttp.ClientResponse`` being raised when a server responds with a 400+ response such as an HTTP
+403.
+
+If downloading synchronously, exceptions are raised from
+:meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.fetch`. If download in parallel, exceptions
```
[the remark this answers is not in the document]
Comment (bmbouter) [reply docs/plugins/plugin-api/asyncio.rst]: fixed in next push

File: docs/plugins/plugin-api/asyncio.rst (hunk 1 above)
Comment (bmbouter) [reply docs/plugins/plugin-api/asyncio.rst]: fixed in next push

File: docs/plugins/plugin-api/asyncio.rst [hunk 3]
```diff
@@ -59,6 +61,91 @@ the downloader's `run()` method when the download is complete.
...
+.. _custom-download-behavior:
+
+Custom Download Behavior
+------------------------
+
+Custom download behavior is provided by subclassing a downloader and providing a new `run()` method.
+For example you could catch a specific error code like a 404 and try another mirror if your
+downloader knew of several mirrors. Here is an `example of that <http://example.com/>`_ in code.
+
+A custom downloader can be given as the downloader to use for a given procol using the
```
[the remark this answers is not in the document]
Comment (bmbouter) [reply docs/plugins/plugin-api/asyncio.rst]: fixed in next push

File: docs/plugins/plugin-api/asyncio.rst [hunk 4]
```diff
@@ -59,6 +61,88 @@ the downloader's `run()` method when the download is complete.
...
+    All :class:`~pulpcore.plugin.download.asyncio.HttpDownload` downloaders produced by the same
+    importer instance share an `aiohttp` session, which provides a connection pool, connection
+    reusage and keep-alives shared across all downloaders produced by a single importer.
+
+.. _exception-handling:
+
+Exception Handling
+------------------
+
+All downloaders are expected to handle recoverable errors automatically. When an unrecoverable error
```
[the remark this answers is not in the document]
Comment (bmbouter) [reply docs/plugins/plugin-api/asyncio.rst]: fixed, good call

**Pull request merged**